src/main/scala/chisel3/testers/TesterDriver.scala

```diff
@@ -2,23 +2,76 @@
 
 package chisel3.testers
 
-import chisel3._
 import java.io._
 
-import chisel3.aop.Aspect
+import chisel3._
 import chisel3.experimental.RunFirrtlTransform
 import chisel3.stage.phases.AspectPhase
-import chisel3.stage.{ChiselCircuitAnnotation, ChiselStage, DesignAnnotation}
-import firrtl.{Driver => _, _}
+import chisel3.stage.{ChiselCircuitAnnotation, ChiselStage, DesignAnnotation, NoRunFirrtlCompilerAnnotation}
+import firrtl.annotations.NoTargetAnnotation
+import firrtl.options.Unserializable
 import firrtl.transforms.BlackBoxSourceHelper.writeResourceToDirectory
+import firrtl.{Driver => _, _}
+import treadle.executable.StopException
+import treadle.stage.TreadleTesterPhase
+import treadle.{CallResetAtStartupAnnotation, TreadleTesterAnnotation, WriteVcdAnnotation}
 
+//scalastyle:off magic.number method.length
 object TesterDriver extends BackendCompilationUtilities {
+  var MaxTreadleCycles = 10000L
+
+  trait Backend extends NoTargetAnnotation with Unserializable
+  case object VerilatorBackend extends Backend
+  case object TreadleBackend extends Backend
+  case object NoBackend extends Backend
+
+  /*
+  Currently the only mechanism for running with the Treadle backend is to edit this
+  statement locally.
+   */
+  val defaultBackend: Backend = VerilatorBackend
+
+  /** Use this to force a test to be run only with backends that are restricted to verilator backend
+    */
+  def verilatorOnly: AnnotationSeq = {
+    if (defaultBackend == TreadleBackend) {
+      Seq(NoBackend)
+    } else {
+      Seq(defaultBackend)
+    }
+  }
+
+  /** For use with modules that should successfully be elaborated by the
+    * frontend, and which can be turned into executables with assertions. */
+  def execute(t:                    () => BasicTester,
+              additionalVResources: Seq[String] = Seq(),
+              annotations:          AnnotationSeq = Seq(),
+              nameHint:             Option[String] = None): Boolean = {
+
+    val backendAnnotations = annotations.collect { case anno: Backend => anno }
+    val backendAnnotation = if (backendAnnotations.length == 1) {
```

Comment:
Random thought: would it be cleaner (or possible) to do a list match? - take the filtered / collected list and match against single-element list/seq/etc cases?